homeassistant/components/fireservicerota/__init__.py

```diff
@@ -0,0 +1,202 @@
+    """Unload FireServiceRota config entry."""
+    hass.data[DOMAIN][entry.entry_id].stop_ws_listener()
+
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in SUPPORTED_PLATFORMS
+            ]
+        )
+    )
+
+    if unload_ok:
+        del hass.data[DOMAIN][entry.entry_id]
+
+    return unload_ok
+
+
+class FSRDataUpdateCoordinator(DataUpdateCoordinator):
+    """Getting the latest data from fireservicerota."""
+
+    def __init__(self, hass, entry):
+        """Initialize the data object."""
+        self._hass = hass
+        self._entry = entry
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_method=self.async_update,
+            update_interval=MIN_TIME_BETWEEN_UPDATES,
+        )
+
+        self._url = entry.data[CONF_URL]
+        self._tokens = entry.data[CONF_TOKEN]
+
+        self.token_refresh_failure = False
+        self.incident_data = None
+        self.incident_id = None
+
+        self.fsr_avail = FireServiceRota(
+            base_url=f"https://{self._url}", token_info=self._tokens
+        )
+
+        self.fsr_incidents = FireServiceRotaIncidents(on_incident=self.on_incident)
+
+        self.start_ws_listener()
+
+    def construct_url(self) -> str:
+        """Return url with latest config values."""
+        return WSS_BWRURL.format(
+            self._entry.data[CONF_URL], self._entry.data[CONF_TOKEN]["access_token"]
+        )
+
+    def on_incident(self, data) -> None:
+        """Update the current data."""
+        _LOGGER.debug("Got data from websocket listener: %s", data)
+        self.incident_data = data
+        dispatcher_send(self._hass, f"{DOMAIN}_{self._entry.entry_id}_update")
+
+    def start_ws_listener(self) -> None:
+        """Start the websocket listener."""
+        _LOGGER.debug("Starting incidents listener")
+        self.fsr_incidents.start(self.construct_url())
+
+    def stop_ws_listener(self) -> None:
+        """Stop the websocket listener."""
+        _LOGGER.debug("Stopping incidents listener")
+        self.fsr_incidents.stop()
+
+    async def update_call(self, func, *args):
+        """Perform update call and return data."""
+        if self.token_refresh_failure:
+            return
+
+        try:
+            return await self._hass.async_add_executor_job(func, *args)
+        except (ExpiredTokenError, InvalidTokenError):
+            if await self.async_refresh_tokens():
+                return await self._hass.async_add_executor_job(func, *args)
+
+    async def async_update(self) -> None:
+        """Get the latest availability data."""
+        _LOGGER.debug("Updating availability data")
+
+        return await self.update_call(
+            self.fsr_avail.get_availability, str(self._hass.config.time_zone)
+        )
+
+    async def async_response_update(self) -> object:
```

Comment:
What is so confusing about this class is that it does 3 things. I think that it should be split out to keep it more clear what's going on:

 - Auth token handling
 - Websocket listening
 - Rest API fetching

Reply:
Yes I agree, even though the code is heavenly optimized and compact, it's still a bit confusing, will do..

Reply:
I have put the websocket and oauth code in their own class now, struggled with oauth code since I need to reference the coordinator early in setup, (to call self.fsr... and self.token_failure... but couldn't use hass.data[DOMAIN][entry.entry_id] for it somehow.
I called the class with a 'self' argument from the coordinator now, it works, but is this allowed??
What do I overlook...

Reply:
It's easier if it's turned around. The auth class shouldn't know how the auth is used, just provide auth tokens, refresh if necessary.

The WebSocket coordinator will take the auth class in their constructor.

Reply:
I have done all changes, and tested them, the normal token refresh and the re-auth config flow is working great. System load is none. 

I have tested with a test FireServiceRota account with emergency events every x seconds, as a stress test. Normally there are only 1 or a few per day for a firefighter.
![image](https://user-images.githubusercontent.com/5447161/99664024-cfb9f200-2a67-11eb-8cc7-0baa3baaffd4.png)

Thank you so far, learned some things along the way, let me know if there is still something to change/add. But it seems it's ready. 
BTW: The CI checks are failing because of something unrelated I think.

Reply:
@balloob Mature enough to merge, or still some work to be done? Eager to open the two PR's for adding switch and binary_sensor.